File: widgets/side_panel.py
"""
右侧辅助面板组件。
包含三个卡片：当前操作提示、标定注意事项、数据文件预览。
"""

from PyQt5.QtCore import Qt
from PyQt5.QtWidgets import (
    QWidget, QVBoxLayout, QHBoxLayout, QFrame, QLabel, QTextEdit, QPushButton,
    QSizePolicy,
)

import resources as res


class SidePanel(QWidget):
    """右侧辅助面板"""

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setMinimumWidth(280)
        self.setSizePolicy(QSizePolicy.Fixed, QSizePolicy.Expanding)

        layout = QVBoxLayout(self)
        layout.setContentsMargins(0, 0, 24, 0)
        layout.setSpacing(12)

        self._tips_card = TipsCard()
        self._notes_card = NotesCard()
        self._preview_card = FilePreviewCard()

        layout.addWidget(self._tips_card)
        layout.addWidget(self._notes_card)
        layout.addWidget(self._preview_card, 1)

    def set_tip(self, text, is_error=False):
        self._tips_card.set_tip(text, is_error)

    def update_file_preview(self, eye_in_hand, marker_type, records):
        self._preview_card.refresh(eye_in_hand, marker_type, records)


class TipsCard(QFrame):
    """当前操作提示卡片"""

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setStyleSheet(f"""
            TipsCard {{
                background-color: {res.BG_MAIN};
                border: 1px solid {res.BORDER_DEFAULT};
                border-radius: {res.BORDER_RADIUS}px;
            }}
        """)

        layout = QVBoxLayout(self)
        layout.setContentsMargins(0, 0, 0, 0)
        layout.setSpacing(0)

        header = QWidget()
        header.setFixedHeight(36)
        header_layout = QHBoxLayout(header)
        header_layout.setContentsMargins(16, 0, 16, 0)
        header_label = QLabel("\U0001f4a1 当前操作提示")
        header_label.setStyleSheet(
            f"font-size: {res.FONT_H2}px; font-weight: 600; "
            f"color: {res.PRIMARY}; border: none;"
        )
        header_layout.addWidget(header_label)
        header_layout.addStretch()
        layout.addWidget(header)

        self._content = QLabel("请先连接相机，然后选择标定模式")
        self._content.setWordWrap(True)
        self._content.setAlignment(Qt.AlignVCenter)
        self._content.setMinimumHeight(48)
        self._content.setStyleSheet(
            f"font-size: {res.FONT_BODY}px; color: {res.TEXT_BODY}; "
            f"border: none; padding: 0 16px 12px 16px;"
        )
        layout.addWidget(self._content)

    def set_tip(self, text, is_error=False):
        self._content.setText(text)
        if is_error:
            self._content.setStyleSheet(
                f"font-size: {res.FONT_BODY}px; color: {res.ERROR}; "
                f"border: none; padding: 0 16px 12px 16px;"
            )
            self.setStyleSheet(f"""
                TipsCard {{
                    background-color: {res.ERROR_BG};
                    border: 1px solid {res.ERROR};
                    border-radius: {res.BORDER_RADIUS}px;
                }}
            """)
        else:
            self._content.setStyleSheet(
                f"font-size: {res.FONT_BODY}px; color: {res.TEXT_BODY}; "
                f"border: none; padding: 0 16px 12px 16px;"
            )
            self.setStyleSheet(f"""
                TipsCard {{
                    background-color: {res.BG_MAIN};
                    border: 1px solid {res.BORDER_DEFAULT};
                    border-radius: {res.BORDER_RADIUS}px;
                }}
            """)


class NotesCard(QFrame):
    """标定注意事项卡片"""

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setStyleSheet(f"""
            NotesCard {{
                background-color: {res.BG_MAIN};
                border: 1px solid {res.BORDER_DEFAULT};
                border-radius: {res.BORDER_RADIUS}px;
            }}
        """)

        layout = QVBoxLayout(self)
        layout.setContentsMargins(0, 0, 0, 0)
        layout.setSpacing(0)

        title_bar = QWidget()
        title_bar.setFixedHeight(36)
        title_bar.setStyleSheet(f"""
            background-color: {res.PRIMARY};
            border-top-left-radius: {res.BORDER_RADIUS}px;
            border-top-right-radius: {res.BORDER_RADIUS}px;
        """)
        title_layout = QHBoxLayout(title_bar)
        title_layout.setContentsMargins(12, 0, 12, 0)
        title_label = QLabel("\U0001f4cb 标定注意事项")
        title_label.setStyleSheet(
            f"font-size: {res.FONT_H2}px; font-weight: 600; "
            f"color: #FFFFFF; border: none;"
        )
        title_layout.addWidget(title_label)
        title_layout.addStretch()
        layout.addWidget(title_bar)

        content = QTextEdit()
        content.setReadOnly(True)
        content.setStyleSheet(f"""
            QTextEdit {{
                border: none;
                font-size: {res.FONT_BODY}px;
                color: {res.TEXT_BODY};
                padding: 16px;
                background-color: {res.BG_MAIN};
                border-bottom-left-radius: {res.BORDER_RADIUS}px;
                border-bottom-right-radius: {res.BORDER_RADIUS}px;
            }}
        """)
        content.setHtml("""
            <p style='line-height: 22px;'>
            <b style='color: #FAAD14;'>⚠️ 标定过程中严禁移动相机和机器人基座</b><br><br>
            • 标定板应尽量充满相机视野的不同区域<br>
            • 建议采集 15-20 组不同位姿的数据<br>
            • 位姿应包含不同的角度和距离<br>
            • 机器人动作幅度越大，标定结果越稳定<br>
            • 确保每次拍照时标定板清晰可见
            </p>
        """)
        layout.addWidget(content)


class FilePreviewCard(QFrame):
    """数据文件预览卡片 — 实时展示已保存的 txt 文件内容"""

    _FILE_SPECS = [
        ("camera_target", "cameraCapturePointXyz.txt", "相机目标点"),
        ("robot_pose", "cameraCaptureRobotPose.txt", "机器人位姿"),
        ("robot_target", "tcp.txt", "机器人目标点"),
    ]

    _MARKER_FILE = ("pose.txt", "机器人拍照位姿")

    def __init__(self, parent=None):
        super().__init__(parent)
        self._eye_in_hand = True
        self._marker_type = True
        self._records = []
        self._active_key = None
        self._tab_btns = {}  # key → QPushButton

        self.setStyleSheet(f"""
            FilePreviewCard {{
                background-color: {res.BG_MAIN};
                border: 1px solid {res.BORDER_DEFAULT};
                border-radius: {res.BORDER_RADIUS}px;
            }}
        """)

        layout = QVBoxLayout(self)
        layout.setContentsMargins(0, 0, 0, 0)
        layout.setSpacing(0)

        # 标题栏（小号字体）
        header = QWidget()
        header.setFixedHeight(30)
        header_layout = QHBoxLayout(header)
        header_layout.setContentsMargins(16, 0, 8, 0)

        title = QLabel("\U0001f4c2 数据文件预览")
        title.setStyleSheet(
            f"font-size: {res.FONT_BODY}px; font-weight: 600; "
            f"color: {res.TEXT_TITLE}; border: none;"
        )
        header_layout.addWidget(title)
        header_layout.addStretch()

        # marker 模式：静态标签
        self._marker_label = QLabel("pose.txt")
        self._marker_label.setStyleSheet(
            f"font-size: {res.FONT_HINT}px; color: {res.TEXT_HINT}; border: none;"
        )
        header_layout.addWidget(self._marker_label)

        layout.addWidget(header)

        # TCP 模式：文件切换标签栏（独立一行，有足够空间）
        self._tab_bar = QWidget()
        tab_bar_layout = QHBoxLayout(self._tab_bar)
        tab_bar_layout.setContentsMargins(12, 2, 12, 4)
        tab_bar_layout.setSpacing(6)

        tab_style = f"""
            QPushButton {{
                background-color: {res.BG_CARD};
                color: {res.TEXT_BODY};
                border: 1px solid {res.BORDER_DEFAULT};
                border-radius: 4px;
                font-size: {res.FONT_HINT}px;
                padding: 4px 10px;
            }}
            QPushButton:checked {{
                background-color: {res.PRIMARY};
                color: #FFFFFF;
                border-color: {res.PRIMARY_CLICK};
                border-bottom: 2px solid {res.PRIMARY_CLICK};
            }}
            QPushButton:hover:!checked {{
                background-color: {res.PRIMARY_LIGHT};
                border-color: {res.PRIMARY};
            }}
        """
        for key, filename, label in self._FILE_SPECS:
            btn = QPushButton(label)
            btn.setCheckable(True)
            btn.setStyleSheet(tab_style)
            btn.setCursor(Qt.PointingHandCursor)
            btn.clicked.connect(self._on_tab_changed)
            self._tab_btns[key] = btn
            tab_bar_layout.addWidget(btn)
        tab_bar_layout.addStretch()
        layout.addWidget(self._tab_bar)

        # 文本预览区
        self._text = QTextEdit()
        self._text.setReadOnly(True)
        self._text.setStyleSheet(f"""
            QTextEdit {{
                border: none;
                font-size: {res.FONT_HINT}px;
                font-family: Consolas, "Courier New", monospace;
                color: {res.TEXT_BODY};
                padding: 8px 12px;
                background-color: {res.BG_MAIN};
                border-bottom-left-radius: {res.BORDER_RADIUS}px;
                border-bottom-right-radius: {res.BORDER_RADIUS}px;
            }}
        """)
        layout.addWidget(self._text)

    def refresh(self, eye_in_hand, marker_type, records):
        self._eye_in_hand = eye_in_hand
        self._marker_type = marker_type
        self._records = records

        if marker_type:
            self._tab_bar.setVisible(False)
            self._marker_label.setVisible(True)
            self._show_marker_content()
        else:
            self._marker_label.setVisible(False)
            self._tab_bar.setVisible(True)
            self._update_tab_visibility()
            # 默认选中第一个可用标签
            available = self._available_keys()
            if self._active_key not in available:
                self._select_tab(available[0] if available else None)

    def _available_keys(self):
        keys = ["camera_target", "robot_target"]
        if self._eye_in_hand:
            keys.insert(1, "robot_pose")
        return keys

    def _update_tab_visibility(self):
        available = set(self._available_keys())
        for key, btn in self._tab_btns.items():
            btn.setVisible(key in available)

    def _select_tab(self, key):
        if key is None:
            return
        self._active_key = key
        for k, btn in self._tab_btns.items():
            btn.setChecked(k == key)
        self._show_tcp_content(key)

    def _on_tab_changed(self, checked):
        # 找出是哪个按钮发出的信号，通过 sender 或遍历
        for key, btn in self._tab_btns.items():
            if btn.isChecked() and key != self._active_key:
                self._active_key = key
                # 取消其他按钮的选中态
                for k2, b2 in self._tab_btns.items():
                    if k2 != key:
                        b2.setChecked(False)
                self._show_tcp_content(key)
                return
        # 如果所有按钮都未选中（用户点击了已选中的按钮），保持当前选中
        if not any(btn.isChecked() for btn in self._tab_btns.values()):
            # 恢复当前选中按钮的状态
            if self._active_key and self._active_key in self._tab_btns:
                self._tab_btns[self._active_key].setChecked(True)

    def _show_marker_content(self):
        filename, _ = self._MARKER_FILE
        lines = [r.robot_capture_pose for r in self._records]
        self._text.setPlainText(self._format_content(filename, lines))

    def _show_tcp_content(self, key):
        field_map = {
            "camera_target": "camera_target_xyz",
            "robot_pose": "robot_capture_pose",
            "robot_target": "robot_target_xyz",
        }
        field = field_map.get(key, "")
        filename = next((f for k, f, _ in self._FILE_SPECS if k == key), "")
        lines = [getattr(r, field, "") for r in self._records]
        self._text.setPlainText(self._format_content(filename, lines))

    @staticmethod
    def _format_content(filename, lines):
        if not lines or all(not s.strip() for s in lines):
            return f"# {filename}\n(暂无数据)"
        return f"# {filename}\n" + "\n".join(str(s) for s in lines)

File: src/ui/SidePanel.cpp
#include "ui/SidePanel.h"
#include "ui/Theme.h"
#include "models/CaptureRecord.h"
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QScrollArea>
#include <QScrollBar>
#include <QFormLayout>
#include <sstream>

SidePanel::SidePanel(QWidget* parent)
    : QWidget(parent)
{
    setFixedWidth(408);
    setStyleSheet(QStringLiteral("background-color: %1;").arg(Theme::BG_MAIN));

    auto* mainLayout = new QVBoxLayout(this);
    mainLayout->setContentsMargins(0, 0, 24, 0);
    mainLayout->setSpacing(12);

    // ── Top area: merged "操作日志" (notes + tips + pose guide + quality) ──
    // Elastic (shares vertical space equally with the file preview card
    // below); append-only log with a scrollbar so users can scroll back to
    // earlier entries while new ones auto-scroll into view. On startup this
    // defaults to the former "标定注意事项" content, tagged [注意事项].
    m_timelineCard = new QFrame(this);
    m_timelineCard->setStyleSheet(QStringLiteral("background: %1; border: 1px solid %2; border-radius: %3px;")
                                  .arg(Theme::BG_MAIN).arg(Theme::BORDER_DEFAULT).arg(Theme::BORDER_RADIUS));
    auto* tlLayout = new QVBoxLayout(m_timelineCard);
    tlLayout->setContentsMargins(12, 10, 12, 10);
    tlLayout->setSpacing(6);

    auto* tlHeader = new QHBoxLayout();
    auto* tlTitle = new QLabel(QStringLiteral("操作日志"), m_timelineCard);
    tlTitle->setStyleSheet(QStringLiteral("font-size: %1px; font-weight: 600; color: %2; border: none;")
                           .arg(Theme::FONT_BODY).arg(Theme::PRIMARY));
    tlHeader->addWidget(tlTitle);
    tlHeader->addStretch();
    m_btnRunQuality = new QPushButton(QStringLiteral("运行质检"), m_timelineCard);
    m_btnRunQuality->setFixedHeight(26);
    m_btnRunQuality->setCursor(Qt::PointingHandCursor);
    m_btnRunQuality->setStyleSheet(QStringLiteral(R"(
        QPushButton { color: %1; background: %2; border: 1px solid %1; border-radius: 4px;
                      font-size: %3px; padding: 0 10px; }
        QPushButton:hover { background: %4; }
    )").arg(Theme::PRIMARY).arg(Theme::BG_MAIN).arg(Theme::FONT_HINT).arg(Theme::PRIMARY_LIGHT));
    tlHeader->addWidget(m_btnRunQuality);
    tlLayout->addLayout(tlHeader);

    m_timeline = new QTextEdit(m_timelineCard);
    m_timeline->setReadOnly(true);
    m_timeline->setStyleSheet(QStringLiteral("background: %1; border: 1px solid %2; border-radius: 4px; "
                                             "font-size: %3px; color: %4;")
                              .arg(Theme::BG_CARD).arg(Theme::BORDER_DEFAULT)
                              .arg(Theme::FONT_HINT).arg(Theme::TEXT_BODY));
    // Default content on first open: the former "标定注意事项" text, tagged
    // [注意事项] so it reads consistently with later [提示]/[姿态引导]/
    // [数据质检] entries appended by setTip()/setPoseGuide()/setQualityReport().
    appendTimeline(QStringLiteral("注意事项"), Theme::TEXT_TITLE,
                   QStringLiteral("标定过程中严禁移动相机和机器人基座"));
    appendTimeline(QStringLiteral("注意事项"), Theme::TEXT_TITLE,
                   QStringLiteral("标定板应尽量充满相机视野的不同区域"));
    appendTimeline(QStringLiteral("注意事项"), Theme::WARNING,
                   QStringLiteral("建议采集 15-20 组不同位姿的数据"));
    appendTimeline(QStringLiteral("注意事项"), Theme::TEXT_TITLE,
                   QStringLiteral("位姿应包含不同的角度和距离"));
    tlLayout->addWidget(m_timeline, 1);
    mainLayout->addWidget(m_timelineCard, 1);

    connect(m_btnRunQuality, &QPushButton::clicked, this, &SidePanel::qualityCheckRequested);

    // ── Bottom area: file preview (shares space equally with 操作日志) ──
    m_previewCard = new QFrame(this);
    m_previewCard->setStyleSheet(QStringLiteral("background: %1; border: 1px solid %2; border-radius: %3px;")
                                 .arg(Theme::BG_MAIN).arg(Theme::BORDER_DEFAULT).arg(Theme::BORDER_RADIUS));
    auto* previewLayout = new QVBoxLayout(m_previewCard);
    previewLayout->setContentsMargins(12, 10, 12, 10);
    previewLayout->setSpacing(6);

    auto* previewTitle = new QLabel(QStringLiteral("文件预览"), m_previewCard);
    previewTitle->setStyleSheet(QStringLiteral("font-size: %1px; font-weight: 600; color: %2; border: none;")
                                .arg(Theme::FONT_BODY).arg(Theme::TEXT_TITLE));
    previewLayout->addWidget(previewTitle);

    // Tab buttons
    auto* tabLayout = new QHBoxLayout();
    tabLayout->setSpacing(4);

    auto makeTab = [&](const QString& text) -> QPushButton* {
        auto* b = new QPushButton(text, m_previewCard);
        b->setCheckable(true);
        b->setFixedHeight(22);
        b->setStyleSheet(QStringLiteral(R"(
            QPushButton { color: %1; background: transparent; border: 1px solid %2;
                          border-radius: 3px; font-size: %3px; padding: 0 8px; }
            QPushButton:hover { border-color: %4; }
            QPushButton:checked { border-color: %4; color: %4; }
        )").arg(Theme::TEXT_HINT).arg(Theme::BORDER_DEFAULT).arg(Theme::FONT_HINT).arg(Theme::PRIMARY));
        return b;
    };

    m_tabCamTarget  = makeTab(QStringLiteral("相机目标点"));
    m_tabRobotPose  = makeTab(QStringLiteral("机器人位姿"));
    m_tabRobotTarget = makeTab(QStringLiteral("机器人TCP点"));
    m_tabCalibResult = makeTab(QStringLiteral("标定结果"));

    tabLayout->addWidget(m_tabCamTarget);
    tabLayout->addWidget(m_tabRobotPose);
    tabLayout->addWidget(m_tabRobotTarget);
    tabLayout->addWidget(m_tabCalibResult);
    tabLayout->addStretch();
    previewLayout->addLayout(tabLayout);

    m_previewText = new QTextEdit(m_previewCard);
    m_previewText->setReadOnly(true);
    m_previewText->setStyleSheet(QStringLiteral("background: %1; border: 1px solid %2; border-radius: 4px; "
                                                 "font-size: %3px; color: %4; font-family: Consolas, monospace;")
                                 .arg(Theme::BG_CARD).arg(Theme::BORDER_DEFAULT)
                                 .arg(Theme::FONT_HINT).arg(Theme::TEXT_HINT));
    previewLayout->addWidget(m_previewText, 1);

    mainLayout->addWidget(m_previewCard, 1);

    // Tab connections
    m_tabCamTarget->setChecked(true);
    m_activeTab = "cam_target";
    auto onTab = [this](const QString& key) {
        if (m_activeTab == key) return;
        m_activeTab = key;
        m_tabCamTarget->setChecked(key == "cam_target");
        m_tabRobotPose->setChecked(key == "robot_pose");
        m_tabRobotTarget->setChecked(key == "robot_target");
        m_tabCalibResult->setChecked(key == "calib_result");
        if (key == "calib_result")
            showCalibrationResult();
        else if (!m_markerType)
            showTcpPreview(m_activeTab, m_eyeInHand, m_cachedRecords);
        else
            showMarkerPreview(m_cachedRecords);
    };
    connect(m_tabCamTarget, &QPushButton::clicked, this, [onTab]() { onTab("cam_target"); });
    connect(m_tabRobotPose, &QPushButton::clicked, this, [onTab]() { onTab("robot_pose"); });
    connect(m_tabRobotTarget, &QPushButton::clicked, this, [onTab]() { onTab("robot_target"); });
    connect(m_tabCalibResult, &QPushButton::clicked, this, [onTab]() { onTab("calib_result"); });
}

void SidePanel::appendTimeline(const QString& label, const QString& colorHex, const QString& text)
{
    const QString html = QStringLiteral(
        "<div style='margin-bottom:6px;'>"
        "<span style='color:%1; font-weight:600;'>[%2]</span> "
        "<span style='color:%3;'>%4</span>"
        "</div>")
        .arg(colorHex).arg(label).arg(Theme::TEXT_BODY).arg(text.toHtmlEscaped());
    m_timeline->append(html);
    QScrollBar* bar = m_timeline->verticalScrollBar();
    if (bar)
        bar->setValue(bar->maximum());
}

void SidePanel::setTip(const QString& text, bool isError)
{
    appendTimeline(QStringLiteral("提示"), isError ? Theme::ERROR : Theme::PRIMARY, text);
}

void SidePanel::setCalibrationResult(const QString& text)
{
    m_calibResult = text;
    if (m_activeTab == QStringLiteral("calib_result"))
        showCalibrationResult();
}

void SidePanel::setPoseGuide(const QString& text, bool warning)
{
    appendTimeline(QStringLiteral("姿态引导"), warning ? Theme::WARNING : Theme::TEXT_TITLE, text);
}

void SidePanel::setQualityReport(const QString& html)
{
    appendTimeline(QStringLiteral("数据质检"), Theme::TEXT_TITLE, html);
}

void SidePanel::updateFilePreview(bool eyeInHand, bool markerType,
                                  const std::vector<CaptureRecord>& records)
{
    m_eyeInHand = eyeInHand;
    m_markerType = markerType;
    m_cachedRecords = records;

    m_tabCamTarget->show();
    m_tabRobotTarget->setVisible(!markerType);
    m_tabRobotPose->setVisible(!(!eyeInHand && !markerType));
    m_tabCalibResult->show();

    if (m_activeTab == QStringLiteral("calib_result")) {
        showCalibrationResult();
    } else if (markerType) {
        showMarkerPreview(records);
    } else {
        showTcpPreview(m_activeTab, eyeInHand, records);
    }
}

void SidePanel::showCalibrationResult()
{
    m_previewText->setText(m_calibResult.isEmpty()
                               ? QStringLiteral("（尚未计算标定结果）")
                               : m_calibResult);
}

void SidePanel::showMarkerPreview(const std::vector<CaptureRecord>& records)
{
    std::ostringstream oss;
    for (const auto& r : records) {
        oss << r.robotCapturePose.toStdString() << "\n";
    }
    m_previewText->setText(QString::fromStdString(oss.str()));
}

void SidePanel::showTcpPreview(const QString& key, bool eyeInHand,
                               const std::vector<CaptureRecord>& records)
{
    std::ostringstream oss;
    for (const auto& r : records) {
        if (key == "cam_target")
            oss << r.cameraTargetXyz.toStdString() << "\n";
        else if (key == "robot_pose" && eyeInHand)
            oss << r.robotCapturePose.toStdString() << "\n";
        else if (key == "robot_target")
            oss << r.robotTargetXyz.toStdString() << "\n";
    }
    m_previewText->setText(QString::fromStdString(oss.str()));
}
